### ISSUE TITLE
Using Custom Converters

### ISSUE THREAD

**Issue by williamoanta** [mrbandler/FsFirestore#4]
I was trying to implement in F# Custom Converters as described [here](https://googleapis.github.io/google-cloud-dotnet/docs/Google.Cloud.Firestore/datamodel.html#custom-converters), but I got into the problem of F#'s files ordering.

First, the type that will be converted requires such an attribute, which needs to reference the ConverterClass - the actual class that does the conversion:
```
[<FirestoreDataAttribute(ConverterType=typeof<ConverterClass>)>]
type TypeClass = 
```

At the same time the ConverterClass needs to implement the following interface based on the Firestore implementation:

```
type IFirestoreConverter<'T> =
        abstract member ToFirestore: 'T -> Object
        abstract member FromFirestore: Object -> 'T
```

Also in the implementation of the interface, we need to use the TypeClass when we implement the two methods to replace the Generic T.

The main problem is that one of the two classes (ConverterClass or TypeClass) will be defined before the other, so then the latter cannot access the former, hence the code does not run.

I haven't seen any implementation of custom converters in the library, and I wanted to ask if there is any solution to the above.
Also, are there any plans to integrate this in the library?

Thank you!

**Comment by mrbandler** [comment]
Hi @oantawilliam, sorry for the late response.

I am currently not able to take a look at it, but I will check with a local example first thing in the morning.

Also, thank you for the question, never needed custom converters but they seem very handy. Since I am currently planning a complete rewrite, as the backing C# library is now out of beta (which it wasn't back when I started the project) I'll try to add a functional friendly way for custom converters.

**Comment by mrbandler** [comment]
After a lot of trial and error, I come to the conclusion that custom converters are currently not possible within an F# setting.

This was my test:
```fsharp
open FsFirestore.Firestore
open FsFirestore.Types
open Google.Cloud.Firestore

/// Actual player ID implementation.
type PlayerId (id: string) =
        let id = id
        member this.Id with get () = id

/// Player ID converter implementation.
type PlayerIdConverter () =
       interface IFirestoreConverter<PlayerId> with
       
       member this.ToFirestore value = value.Id :> obj
       
       member this.FromFirestore value =
               match value with
               | :? string as id -> PlayerId id
               | _ -> PlayerId ""

/// Let's try with inheritance and apply the attribute here.              
[<FirestoreData(ConverterType = typeof<PlayerIdConverter>)>]
type PlayerIdC (id) =
        inherit PlayerId (id)
        
/// Player ID usage.
[<FirestoreData>]
type Game () =
    inherit FirestoreDocument()
    
    [<FirestoreProperty>]
    member val PlayerA = PlayerIdC "0"
    
    [<FirestoreProperty>]
    member val PlayerB = PlayerIdC "1"

[<EntryPoint>]
let main argv =
    let didConnect = connectToFirestore "./GCP.json" 
    printfn "Connected: %b" didConnect
    
    let game = Game ()
    let ref = addDocument "games" (Some "42") game
    
    0
```

Sadly this doesn't work as the reflection mechanism from the Google Cloud Firestore library expects the `PlayerId` type to be used. The converter creator and lookup then tries to create a `IFirestoreConverter<PlayerIdC>` converter which has not been defined.

Possible solutions are, deep type inspection while creating converters on the Google Cloud Firestore library side, meaning a complete type inspection of the inheritance chain all the way to the bottom type (which might cause performance issues if huge inheritance chains are present in the code base) or implementation of something like [this](https://googleapis.dev/dotnet/Google.Cloud.Firestore/latest/datamodel.html#converter-registries) on the FsFirestore end:

```fsharp
FirestoreDb db = new FirestoreDbBuilder
{
    ProjectId = projectId,
    ConverterRegistry = new ConverterRegistry
    {
        new GuidConverter()
    }
}.Build();
```

The latter would probably be the fastest and easiest one to implement. I will try to push the feature update tomorrow evening and let you know in this thread.

**Comment by mrbandler** [comment]
Alrighty, I just published v1.3.0.

I added the ability to connect to your Firestore instance given a DB builder.
This, in general, adds more flexibility to the library.

I also added an example of how to use the new connection functions with a builder, one of the examples even shows how to use a custom converter and how to register it. [Here](https://github.com/mrbandler/FsFirestore#3-service-account-json-path--db-builder) is a direct link.

For easier reference in the issue I also add the example code here:

```fsharp
open FsFirestore.Firestore
open Google.Cloud.Firestore

// Player ID.
type PlayerId (id: string) =
    let id = id
    member this.Id with get () = id

// Custom player ID converter.
type PlayerIdConverter () =
    interface IFirestoreConverter<PlayerId> with

    member this.ToFirestore value = value.Id :> obj

    member this.FromFirestore value =
        match value with
        | :? string as id -> PlayerId id
        | _ -> PlayerId ""

// Use custom type with custom converter.
[<FirestoreData>]
type Game () =
    inherit FirestoreDocument()

    [<FirestoreProperty>]
    member val PlayerA = PlayerId "0"

    [<FirestoreProperty>]
    member val PlayerB = PlayerId "1"

// connect :: Unit -> Boolean
let connect () =
    let converter = PlayerIdConverter ()
    let registry = ConverterRegistry ()
    registry.Add converter

    let builder = FirestoreDbBuilder ()
    builder.ConverterRegistry <- registry

    builder |> connectToFirestoreWithBuilder "./path/to/your/service_account.json"
```
